lib/bolt/plugin/module.rb

```diff
@@ -0,0 +1,238 @@
+# frozen_string_literal: true
+
+require 'bolt/task/run'
+
+module Bolt
+  class Plugin
+    class Module
+      class InvalidPluginData < Bolt::Plugin::PluginError
+        def initialize(plugin, msg)
+          msg = "Invalid Plugin Data for #{plugin}: #{msg}"
+          super(msg, 'bolt/invalid-plugin-data')
+        end
+      end
+
+      def self.load(name, modules, opts)
+        mod = modules[name]
+        if mod&.plugin?
+          opts[:mod] = mod
+          plugin = Bolt::Plugin::Module.new(opts)
+          plugin.setup
+          plugin
+        else
+          raise PluginError::Unknown, name
+        end
+      end
+
+      attr_reader :config
+
+      def initialize(mod:, context:, config:, **_opts)
+        @module = mod
+        @config = config
+        @context = context
+      end
+
+      # This method interacts with the module on disk so it's separate from initialize
+      def setup
+        @data = load_data
+        @config_schema = process_schema(@data['config'] || {})
+
+        @hook_map = find_hooks(@data['hooks'] || {})
+
+        validate_config(@config, @config_schema)
+      end
+
+      def name
+        @module.name
+      end
+
+      def hooks
+        (@hook_map.keys + [:validate_resolve_reference]).uniq
+      end
+
+      def config?
+        @data.include?('config') && !@data['config'].empty?
+      end
+
+      def load_data
+        JSON.parse(File.read(@module.plugin_data_file))
+      rescue JSON::ParserError => e
+        raise InvalidPluginData.new(e.message, name)
+      end
+
+      def process_schema(schema)
+        raise InvalidPluginData.new('config specification is not an object', name) unless schema.is_a?(Hash)
+        schema.each do |key, val|
+          unless key =~ /\w+/ # TODO: Where is the parameter regex?
```

Comment:
Is it the one in puppet-specifications? `\A[a-z][a-z0-9_]*\z` https://github.com/puppetlabs/puppet-specifications/tree/master/tasks#options

Reply:
Here's where it's enforced: 
 https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/pcore.rb#L51